csp-mixin/signals/virtualMachines.libsonnet

```
@@ -0,0 +1,141 @@
+local commonlib = import 'common-lib/common/main.libsonnet';
+function(this)
+  {
+    local s = self,
+    filteringSelector: this.filteringSelector,
+    groupLabels: this.groupLabels,
+    instanceLabels: this.instanceLabels,
+    aggLevel: 'instance',
+    discoveryMetric: {
+      azuremonitor: 'azure_microsoft_compute_virtualmachines_vmavailabilitymetric_average_count',
+    },
+    signals: {
+      cpuUtilization: {
+        name: 'CPU Utilization average',
+        description: 'The percentage of allocated compute units that are currently in use by the Virtual Machine(s)',
+        type: 'raw',
+        unit: 'percent',
+        sources: {
+          azuremonitor: {
+            expr: 'avg by (resourceName, job, resourceGroup, subscriptionName) (azure_microsoft_compute_virtualmachines_percentage_cpu_average_percent{%(queriesSelector)s})',
+            legendCustomTemplate: '{{resourceName}}',
+          },
+        },
+      },
+
+      availableMemory: {
+        name: 'Available memory',
+        description: 'Amount of physical memory, in bytes, immediately available for allocation to a process or for system use in the Virtual Machine',
+        type: 'raw',
+        unit: 'decbytes',
+        sources: {
+          azuremonitor: {
+            expr: 'sum by (resourceName, job, resourceGroup, subscriptionName) (azure_microsoft_compute_virtualmachines_available_memory_bytes_average_bytes{%(queriesSelector)s})',
+            legendCustomTemplate: '{{resourceName}}',
+          },
+        },
+      },
```

you can avoid using raw everywhere and keep expr clean.
For example:

```suggestion
      availableMemory: {
        name: 'Available memory',
        description: 'Amount of physical memory, in bytes, immediately available for allocation to a process or for system use in the Virtual Machine',
        type: 'gauge
        unit: 'decbytes',
        aggFunction: 'sum',
        sources: {
          azuremonitor: {
            expr: azure_microsoft_compute_virtualmachines_available_memory_bytes_average_bytes{%(queriesSelector)s}',
            legendCustomTemplate: '{{resourceName}}',
            aggKeepLabels: ['resourceName', 'resourceGroup', 'subscriptionName'],
          },
        },
      },
```

this will sum by aggKeepLabels and by dynamically set `groupLabels`, `instanceLabels` like job, instance in the default case. (due to signals group aggLevel=instance)

---

Love it. I never use the other signals **type** because I don't know which information they add to the expression. I will change it in the rest of the queries. thanks